pkg/kube/clients.go

```diff
@@ -0,0 +1,143 @@
+// limitations under the License.
+
+package kube
+
+import (
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/dynamic/dynamicinformer"
+	dynamicfake "k8s.io/client-go/dynamic/fake"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/fake"
+	"k8s.io/client-go/metadata"
+	metadatafake "k8s.io/client-go/metadata/fake"
+	"k8s.io/client-go/metadata/metadatainformer"
+	"k8s.io/client-go/rest"
+)
+
+type Clients struct {
+	kube         kubernetes.Interface
+	kubeInformer informers.SharedInformerFactory
+
+	dynamic         dynamic.Interface
+	dynamicInformer dynamicinformer.DynamicSharedInformerFactory
+
+	metadata         metadata.Interface
+	metadataInformer metadatainformer.SharedInformerFactory
+	// TODO: add istio/kube-go
+	// TODO: add service-apis
+}
+
+const resyncInterval time.Duration = 0
+
+// NewFakeClients builds a Clients made entirely of fakes for testing purposes
+func NewFakeClients() *Clients {
+	var clients Clients
+	clients.kube = fake.NewSimpleClientset()
+	clients.kubeInformer = informers.NewSharedInformerFactory(clients.kube, resyncInterval)
+
+	scheme := runtime.NewScheme()
+	if err := metav1.AddMetaToScheme(scheme); err != nil {
+		panic(err.Error())
+	}
+	clients.metadata = metadatafake.NewSimpleMetadataClient(scheme)
+	clients.metadataInformer = metadatainformer.NewSharedInformerFactory(clients.metadata, resyncInterval)
+
+	clients.dynamic = dynamicfake.NewSimpleDynamicClient(scheme)
+	clients.dynamicInformer = dynamicinformer.NewDynamicSharedInformerFactory(clients.dynamic, resyncInterval)
+
+	return &clients
+
+}
+
+func NewClients(config *rest.Config) (*Clients, error) {
+	var clients Clients
+	var err error
+
+	clients.kube, err = kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+	clients.kubeInformer = informers.NewSharedInformerFactory(clients.kube, resyncInterval)
+
+	clients.metadata, err = metadata.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+	clients.metadataInformer = metadatainformer.NewSharedInformerFactory(clients.metadata, resyncInterval)
+
+	clients.dynamic, err = dynamic.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+	clients.dynamicInformer = dynamicinformer.NewDynamicSharedInformerFactory(clients.dynamic, resyncInterval)
+
+	return &clients, nil
+}
+
+// NewClientsFromKubeconfig returns a new set of the various API clients required by Istio
+// TODO: add watchedNamespaces
+func NewClientsFromKubeconfig(kubeconfig string, context string, fns ...func(*rest.Config)) (*Clients, error) {
+	config, err := BuildClientConfig(kubeconfig, context)
+	if err != nil {
+		return nil, err
+	}
+	for _, fn := range fns {
+		fn(config)
+	}
+	return NewClients(config)
+}
+
+// Client returns the kube Kubernetes kube.
+func (c *Clients) Kube() kubernetes.Interface {
```

Comment:
This looks suspiciously like https://github.com/istio/istio/blob/master/pkg/kube/client.go.

I think we should be able to have this extend the other to add informers, etc. Maybe this should be called something like `InformerClient`?

Reply:
I really don't want to pass around a client that has things like `AllDiscoveryDo` to the core of pilot -  that is a mess. Rather, I think that one should actually depend on this one and provide additional functionality. But certainly not in this PR, I am trying to separate the creation of this new interface and the use into a different pr - I just did the kube registry in this PR to prove its viable